File: ProductDevelopment/VisualStudioAddIns/LICode/Code/LICodeHandler.cs
using System;
using EnvDTE;
using EnvDTE80;
using System.Diagnostics;
using System.Text;
using System.Text.RegularExpressions;
using System.IO;
using Extract.SourceControl;
using System.Globalization;

namespace LICode
{
    /// <summary>
    /// The type of the location identifier.
    /// </summary>
    public enum LIType
    {
        /// <summary>
        /// Exception location identifier
        /// </summary>
        Exception = 0,

        /// <summary>
        /// Method location identifier
        /// </summary>
        Method = 1
    }

    // Handles the retrieval and modification of the Location Identifier dat files
    class LICodeHandler : IDisposable
    {
        // Location identifier number format string (five digits)
        const string LI_NUM_FORMAT = "D5";

        // Location identifier dat file data type
        class LIDatFileData
        {
            // Source safe path to the dat file (eg. $\Engineering\file.ext)
            public string vssPath;

            // Source safe item corresponding to dat file
            public ISourceControlItem vssItem;

            // String builder containing lines to append to dat file for new LI codes
            public StringBuilder builder;

            // The prefix of the location identifier code (Exception or Method)
            public string LIPrefix;

            // The location identifier number (eg. 12345 in ELI12345)
            public int LINum = -1;

            // Source safe path and location identifier prefix must be specified at instantiation
            public LIDatFileData(string newVssPath, string newLIPrefix)
            {
                vssPath = newVssPath;
                LIPrefix = newLIPrefix;
            }
        }

        // Visual SourceSafe database
        ISourceControl vssDB;

        // Array of location identifier dat file data objects
        LIDatFileData[] LIDatFiles = { new LIDatFileData("$/Engineering/ProductDevelopment/Common/UCLIDExceptionLocationIdentifiers.dat", "ELI"), new LIDatFileData("$/Engineering/ProductDevelopment/Common/ExtractMethodLocationIdentifiers.dat", "MLI") };

        public LICodeHandler() : this(null)
        {
        }

        public LICodeHandler(ISourceControl vssDatabase)
        {
            vssDB = vssDatabase;
        }

        public string NextELICode
        {
            get
            {
                // open the Visual SourceSafe database if not already opened
                if (vssDB == null)
                {
                    vssDB = SourceControlMethods.OpenSourceControlDatabase();
                }

                // retrieve the Exception dat file if it has not already been retrieved
                RetrieveLIDatFileFromVSS(LIDatFiles[(int)LIType.Exception]);

                return GetNextLICode((int)LIType.Exception);
            }
        }

        public string NextMLICode
        {
            get
            {
                // open the Visual SourceSafe database if not already opened
                if (vssDB == null)
                {
                    vssDB = SourceControlMethods.OpenSourceControlDatabase();
                }

                // retrieve the Method dat file if it has not already been retrieved
                RetrieveLIDatFileFromVSS(LIDatFiles[(int)LIType.Method]);

                return GetNextLICode((int)LIType.Method);
            }
        }

        public string ReplaceLICodesInText(string textToReplace)
        {

            // open the Visual SourceSafe database if not already opened
            if (vssDB == null)
            {
                vssDB = SourceControlMethods.OpenSourceControlDatabase();
            }

            // retrieve both dat files
            foreach (LIDatFileData datFile in LIDatFiles)
            {
                // get the vss item for the LI dat file if not already retrieved
                RetrieveLIDatFileFromVSS(datFile);
            }

            // replace the selected text with new LI codes
            Regex regex = new Regex("\"(M|E)LI\\d+\"", RegexOptions.Compiled);
            return regex.Replace(textToReplace, ReplaceLI);
        }

        public void CommitChanges()
        {
            foreach (LIDatFileData datFile in LIDatFiles)
            {
                // check if the LIBuilder exists
                if (datFile.builder != null)
                {
                    // check if any changes to the LI dat file need to be committed
                    if (datFile.builder.Length > 0)
                    {
                        // add newly added LI codes to the dat file and check it in
                        File.AppendAllText(datFile.vssItem.LocalSpec, datFile.builder.ToString());
                        datFile.vssItem.CheckIn();
                    }
                    else
                    {

                        // undo the checkout on the LI dat file
                        datFile.vssItem.UndoCheckOut();
                    }

                    // reset the values of the current LI number and the LI string builder
                    datFile.LINum = -1;
                    datFile.builder = null;
                    datFile.vssItem = null;
                }
            }
        }

        static int GetLastLICodeInDatFile(string filename)
        {
            // open the LI dat file as a comma-delimited file
            using (StreamReader reader = new StreamReader(filename))
            {
                // Iterate through each line until the end of the file
                string line = null;
                while (!reader.EndOfStream)
                {
                    line = reader.ReadLine();
                }

                // Get the LI number of the last entry
                int index = line.IndexOf(',');
                if (index > 2)
                {
                    string code = line.Substring(3, index - 3);
                    return int.Parse(code, CultureInfo.CurrentCulture);
                }
            }

            return -1;
        }

        string GetNextLICode(int index)
        {
            // increment the LI number
            LIDatFiles[index].LINum += 1;

            // create a new line for this LI code
            LIDatFiles[index].builder.Append(LIDatFiles[index].LIPrefix 
                + LIDatFiles[index].LINum.ToString(LI_NUM_FORMAT, CultureInfo.CurrentCulture) 
                + "," + Environment.UserName + "," + Environment.MachineName + "," 
                + System.DateTime.Now.ToString("MM/dd/yyyy,HH:mm:ss", CultureInfo.InvariantCulture) 
                + ",," + Environment.NewLine);

            // return this LI number
            return "\"" + LIDatFiles[index].LIPrefix 
                + LIDatFiles[index].LINum.ToString(LI_NUM_FORMAT, CultureInfo.CurrentCulture) + "\"";
        }

        string ReplaceLI(Match match)
        {
            // TODO: add progress status

            // get the LI dat file data associated with this match
            int i = (int)(match.Value[1] == 'E' ? 0 : 1);

            // get increment and return the LI code number
            return GetNextLICode(i);
        }

        void RetrieveLIDatFileFromVSS(LIDatFileData datFile)
        {
            if (datFile.LINum == -1)
            {
                datFile.vssItem = SourceControlMethods.CheckOutItemFromPath(vssDB, datFile.vssPath);
                datFile.builder = new StringBuilder();

                // get the last LI code from the dat file
                datFile.LINum = GetLastLICodeInDatFile(datFile.vssItem.LocalSpec);
            }
        }

        #region IDisposable Members

        /// <summary>
        /// Releases all resources used by the <see cref="LICodeHandler"/>.
        /// </summary>
        public void Dispose()
        {
            Dispose(true);
            GC.SuppressFinalize(this);
        }

        /// <overloads>Releases resources used by the <see cref="LICodeHandler"/>.</overloads>
        /// <summary>
        /// Releases all unmanaged resources used by the <see cref="LICodeHandler"/>.
        /// </summary>
        /// <param name="disposing"><see langword="true"/> to release both managed and unmanaged 
        /// resources; <see langword="false"/> to release only unmanaged resources.</param>        
        protected virtual void Dispose(bool disposing)
        {
            if (disposing)
            {
                // Dispose of managed objects
                if (LIDatFiles != null)
                {
                    // undo the check out of the Exception dat file if it is currently held
                    foreach (LIDatFileData datFile in LIDatFiles)
                    {
                        if (datFile != null && datFile.vssItem != null && datFile.vssItem.IsCheckedOutToMe)
                        {
                            datFile.vssItem.UndoCheckOut();
                        }
                    }
                    LIDatFiles = null;
                }
            }

            // Dispose of unmanaged resources
        }

        #endregion IDisposable Members
    }
}

File: ProductDevelopment/Utils/VisualStudioMacros/Common/LICodeHandler.vb
Option Strict On
Option Explicit On
Option Compare Binary
Imports System
Imports EnvDTE
Imports EnvDTE80
Imports System.Diagnostics
Imports System.Text.RegularExpressions

Enum LITypeNum
    ELI = 0
    MLI = 1
End Enum

' Handles the retrieval and modification of the Location Identifier dat files
Class LICodeHandler
    Implements IDisposable

    ' Location identifier number format string (five digits)
    Const LI_NUM_FORMAT As String = "D5"

    ' Location identifier dat file data type
    Private Class LIDatFileData

        ' Source safe path to the dat file (eg. $\Engineering\file.ext)
        Public vssPath As String

        ' Source safe item corresponding to dat file
        Public vssItem As SourceSafeTypeLib.VSSItem

        ' String builder containing lines to append to dat file for new LI codes
        Public builder As Text.StringBuilder

        ' The prefix of the location identifier code (ELI or MLI)
        Public LIPrefix As String

        ' The location identifier number (eg. 12345 in ELI12345)
        Public LINum As Integer = -1

        ' Source safe path and location identifier prefix must be specified at instantiation
        Sub New(ByVal newVssPath As String, ByVal newLIPrefix As String)
            vssPath = newVssPath
            LIPrefix = newLIPrefix
        End Sub
    End Class

    ' Visual SourceSafe database
    Private vssDB As SourceSafeTypeLib.VSSDatabase

    ' Array of location identifier dat file data objects
    Private LIDatFiles() As LIDatFileData = { _
        New LIDatFileData("$/Engineering/ProductDevelopment/Common/UCLIDExceptionLocationIdentifiers.dat", "ELI"), _
        New LIDatFileData("$/Engineering/ProductDevelopment/Common/ExtractMethodLocationIdentifiers.dat", "MLI")}

    ' Whether this object has been disposed
    Protected disposed As Boolean = False

    Sub New()
        ' do nothing of consequence
    End Sub

    Sub New(ByVal vssDatabase As SourceSafeTypeLib.VSSDatabase)

        ' store the VSS database
        vssDB = vssDatabase
    End Sub

    ReadOnly Property NextELICode() As String
        Get
            ' open the Visual SourceSafe database if not already opened
            If vssDB Is Nothing Then
                vssDB = OpenVSSDatabase()
            End If

            ' retrieve the ELI dat file if it has not already been retrieved
            RetrieveLIDatFileFromVSS(LIDatFiles(LITypeNum.ELI))

            Return GetNextLICode(LITypeNum.ELI)
        End Get
    End Property

    ReadOnly Property NextMLICode() As String
        Get
            ' open the Visual SourceSafe database if not already opened
            If vssDB Is Nothing Then
                vssDB = OpenVSSDatabase()
            End If

            ' retrieve the MLI dat file if it has not already been retrieved
            RetrieveLIDatFileFromVSS(LIDatFiles(LITypeNum.MLI))

            Return GetNextLICode(LITypeNum.MLI)
        End Get
    End Property

    Function ReplaceLICodesInText(ByVal textToReplace As String) As String

        ' open the Visual SourceSafe database if not already opened
        If vssDB Is Nothing Then
            vssDB = OpenVSSDatabase()
        End If

        ' retrieve both dat files
        For Each datFile As LIDatFileData In LIDatFiles
            ' get the vss item for the LI dat file if not already retrieved
            RetrieveLIDatFileFromVSS(datFile)
        Next

        ' replace the selected text with new LI codes
        Dim regex As New Regex("""(M|E)LI\d+""")
        Return regex.Replace(textToReplace, AddressOf ReplaceLI)
    End Function

    Sub CommitChanges()

        For Each datFile As LIDatFileData In LIDatFiles

            ' check if the LIBuilder exists
            If datFile.builder IsNot Nothing Then
                With datFile

                    ' check if any changes to the LI dat file need to be committed
                    If .builder.Length > 0 Then

                        ' add newly added LI codes to the dat file and check it in
                        My.Computer.FileSystem.WriteAllText(.vssItem.LocalSpec, .builder.ToString(), True)
                        .vssItem.Checkin()
                    Else

                        ' undo the checkout on the LI dat file
                        .vssItem.UndoCheckout()
                    End If

                    ' reset the values of the current LI number and the LI string builder
                    .LINum = -1
                    .builder = Nothing
                End With
            End If
        Next

    End Sub

    Private Function GetLastLICodeInDatFile(ByVal filename As String) As Integer

        ' open the LI dat file as a comma-delimited file
        Using parser As New FileIO.TextFieldParser(filename)
            With parser
                .TextFieldType = FileIO.FieldType.Delimited
                .Delimiters = New String() {","}

                ' iterate through each entry until the end of the file
                Dim row() As String
                While Not .EndOfData
                    row = .ReadFields()
                End While

                ' get the LI number of the last entry
                Return CInt(row(0).Remove(0, 3))
            End With
        End Using

    End Function

    Private Function GetNextLICode(ByVal index As Integer) As String

        With LIDatFiles(index)

            ' increment the LI number
            .LINum += 1

            ' create a new line for this LI code
            .builder.Append(.LIPrefix & .LINum.ToString(LI_NUM_FORMAT) & "," _
              & Microsoft.VisualBasic.Environ("USERNAME") & "," _
              & Microsoft.VisualBasic.Environ("COMPUTERNAME") & "," _
              & Date.Now.ToString("MM/dd/yyyy,HH:mm:ss") & ",," & vbCrLf)

            ' return this LI number
            Return """" & .LIPrefix & .LINum.ToString(LI_NUM_FORMAT) & """"
        End With

    End Function

    Private Function ReplaceLI(ByVal match As Match) As String

        ' TODO: add progress status

        ' get the LI dat file data associated with this match
        Dim i As Integer = CInt(IIf(match.Value.Chars(1) = "E"c, 0, 1))

        ' get increment and return the LI code number
        Return GetNextLICode(i)

    End Function

    Private Sub RetrieveLIDatFileFromVSS(ByVal datFile As LIDatFileData)
        With datFile
            If .LINum = -1 Then
                .vssItem = CheckoutVSSItemFromVSSPath(vssDB, .vssPath)
                .builder = New Text.StringBuilder

                ' get the last LI code from the dat file
                .LINum = GetLastLICodeInDatFile(.vssItem.LocalSpec)

            End If
        End With
    End Sub

    Protected Overridable Sub Dispose(ByVal disposing As Boolean)
        If Not disposed Then

            ' undo the check out of the ELI dat file if it is currently held
            For Each datFile As LIDatFileData In LIDatFiles
                With datFile
                    If .vssItem IsNot Nothing And .vssItem.IsCheckedOut = SourceSafeTypeLib.VSSFileStatus.VSSFILE_CHECKEDOUT_ME Then
                        .vssItem.UndoCheckout()
                    End If
                End With
            Next

            disposed = True
        End If
    End Sub

    Public Sub Dispose() Implements IDisposable.Dispose
        Dispose(True)
        GC.SuppressFinalize(Me)
    End Sub

    Protected Overrides Sub Finalize()
        Dispose(False)
        MyBase.Finalize()
    End Sub

End Class

